tests/test_cancellations.py

```diff
@@ -0,0 +1,95 @@
+import typing
+
+import anyio
+import pytest
+
+import httpcore
+
+
+class SlowWriteStream(httpcore.AsyncNetworkStream):
+    async def write(
+        self, buffer: bytes, timeout: typing.Optional[float] = None
+    ) -> None:
+        await anyio.sleep(2)
+
+    async def aclose(self) -> None:
+        pass
+
+
+class SlowReadStream(httpcore.AsyncNetworkStream):
+    def __init__(self, buffer: typing.List[bytes]):
+        self._buffer = buffer
+
+    async def write(self, buffer, timeout=None):
+        pass
+
+    async def read(
+        self, max_bytes: int, timeout: typing.Optional[float] = None
+    ) -> bytes:
+        if not self._buffer:
+            await anyio.sleep(2)
+        else:
+            return self._buffer.pop(0)
+
+    async def aclose(self):
+        pass
+
+
+class SlowBackend(httpcore.AsyncNetworkBackend):
+    async def connect_tcp(
+        self,
+        host: str,
+        port: int,
+        timeout: typing.Optional[float] = None,
+        local_address: typing.Optional[str] = None,
+        socket_options: typing.Optional[typing.Iterable[httpcore.SOCKET_OPTION]] = None,
+    ) -> httpcore.AsyncNetworkStream:
+        return SlowWriteStream()
+
+
+@pytest.mark.anyio
+async def test_async_cancellation():
+    async with httpcore.AsyncConnectionPool(network_backend=SlowBackend()) as pool:
+        with anyio.move_on_after(0.001):
+            await pool.request("GET", "http://example.com")
+        assert not pool.connections
+
+
+@pytest.mark.anyio
+async def test_h11_response_closed():
+    origin = httpcore.Origin(b"http", b"example.com", 80)
+    stream = SlowWriteStream()
+    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
+        with anyio.move_on_after(0.001):
+            await conn.request("GET", "http://example.com")
+        assert conn.is_closed()
+
+
+@pytest.mark.anyio
+async def test_h2_response_closed():
+    origin = httpcore.Origin(b"http", b"example.com", 80)
+    stream = SlowWriteStream()
+    async with httpcore.AsyncHTTP2Connection(origin, stream) as conn:
+        with anyio.move_on_after(0.001):
+            await conn.request("GET", "http://example.com")
+        assert conn.is_closed()
+
+
+@pytest.mark.anyio
+async def test_h11_bytestream_cancellation():
+    origin = httpcore.Origin(b"http", b"example.com", 80)
+    stream = SlowReadStream(
+        [
+            b"HTTP/1.1 200 OK\r\n",
+            b"Content-Type: plain/text\r\n",
+            b"Content-Length: 1000\r\n",
+            b"\r\n",
+            b"Hello, world!...",
+        ]
+    )
+    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
+        with anyio.move_on_after(0.001):
+            async with conn.stream("GET", "http://example.com") as resp:
+                async for chunk in resp.aiter_stream():
+                    pass
```

Comment:
Just `conn.request(...)` would be sufficient here.

(We'll still end up calling into `HTTP11ConnectionByteStream.__aiter__`)